### ISSUE TITLE
Add SSL into minetest

### ISSUE THREAD

**Issue by Emojigit** [minetest/minetest#10206]
## Problem

I don't know how minetest keep the data transform secret.
If the password authorize data are not safe, a hacker can easily hack into a player's account.

## Solutions

I think minetest shoud use SSL to keep the datas secret. (That will make the client below 5.4.0 can't link to 5.4.0 servers.)

## Alternatives

Make a client-side mod to let other version that below 5.4.0 know how to link via SSL

## Additional context

SSL version: TLS1.3

**Comment by SmallJoker** [comment]
Passwords are encoded using SRP before they're sent over the network, requiring quite some efforts by the attacker. See also: #1943 -> PR #2620.

DTLS would be a possibility instead of pain UDP. However, the transferred packets are usually very small (< 1 KiB), with unreliable parts mixed into it (player movements). I don't see any need to encrypt the network - this is a game, not a nuclear reactor remote software (which shouldn't exist in the first place ;) ).

**Comment by hecktest** [comment]
Possible close, really?
Chat and form traffic should be encrypted. Maybe meta as well. Anything that involves big blobs of text.

As it is, a MITM attack can be used to impersonate someone with privileges and send commands in their stead, including very dangerous ones like #10163, making this a vector for a server to get compromised. (We have enough botnets on the internet as it is.)

In addition, lots of kids play this game without knowing that their communications can be trivially overheard. By not encrypting the traffic, you are making it very easy for some creep to datamine them. (Researching the victim is usually the first step for kidnappers, traffickers, etc. If you think they wouldn't bother bugging someone's home connection, think again.)

Or maybe somebody lives in a country with a less than stellar human rights situation, and slips up in conversation. One word might be enough to land you in trouble, or put you on some list. Do you think an authoritarian government cares if it happened in a block game?

Finally, if a zero day is found in some more complicated portion of the client environment such as forms, malware can be spread without having to compromise the server itself. Any node along the way can do it. Yikes.

And if any of those things happen, the server owner might have to deal with some unpleasant fallout too, depending on the circumstances.

The question about encrypting communication and securing a public server in 2020 should never be "if" but "how". Don't wait until any of the above things happen, do it now.

**Comment by rubenwardy** [comment]
I consider this low priority at this time. This will probably be needed when server-sent client-side scripting is sent, but it's not required right now. Chat is not privileged, and connection hijacking is not trivial (you need to be on a node between the server and the client) - Minetest is also not that big a target.

Now-a-days, encryption has good hardware support so performance shouldn't be a problem. We can also use the handshake to let the client and server agree on whether encryption should be used, allowing network compatibility to be kept. This would introduce a downgrade attack, but we could use the serverlist to require encryption client-side when connecting. Server-sent client-side mods could also require encryption.

Authentication of the server could be an issue, the best way to do that would be two methods:

- the server list could contain the public key of the server
- the public key of a server could be verified by its domain name, using standard internet CAs

**Comment by hecktest** [comment]
Minetest is not big but in my opinion we shouldn't wait with security until someone dies. It's not like encryption is hard either.
Using the master server as a key authority should be good enough as a default. Third party CAs would require some extra steps to set up, like having a domain in the first place. That part need not be in the initial implementation.
Plaintext connections would still be supported, but with a little string on the loading screen warning you about it, and maybe a padlock icon in the list for secure servers.

And oh yeah, I forgot SSCSM. That _requires_ encryption, not for secrecy but for integrity. You might want to add this issue to the SSCSM project.

**Comment by rubenwardy** [comment]
The idea behind domain certs and CAs would be to allow authentication for servers not on the server list, but that's a much lower priority and that solution doesn't work for all servers as you said

**Comment by hecktest** [comment]
CAs would help but they're a bit more work, hence the lower priority for them. In the meantime, being able to manually inspect the cert would be useful for the unlisted server case.

**Comment by Lejo1** [comment]
Isn't SRP able to create a symmetric session key for encryption?

Problems:
First SRP is still insecure

**Comment by Emojigit** [comment]
> Passwords are encoded using SRP before they're sent over the network, requiring quite some efforts by the attacker. See also: #1943 -> PR #2620.
> 
> DTLS would be a possibility instead of pain UDP. However, the transferred packets are usually very small (< 1 KiB), with unreliable parts mixed into it (player movements). I don't see any need to encrypt the network - this is a game, not a nuclear reactor remote software (which shouldn't exist in the first place ;) ).

I think player move should encrypt because a hacker can move you into lava if player move are not encrypted

**Comment by Emojigit** [comment]
I asked on stack overflow: https://stackoverflow.com/questions/63171893/are-there-any-network-encryption-for-udp-that-support-big-packet

**Comment by Lejo1** [comment]
> I think player move should encrypt because a hacker can move you into lava if player move are not encrypted

That would require the hacker to do a middle man attack which is nothing you just do for killing a player in a game...

Sure there are Udp encryption algorithms.
The Problem is just that encryption/decryption takes time, so a fast one should be used.

**Comment by hecktest** [comment]
The reason movement shouldn't be encrypted is because it's a low latency, high frequency kind of information which must be sent as quickly as possible, and there's nothing sensitive about it. Same goes for plain button inputs.

I do wonder about the possibility of griefing a server with spoofed input packets, re-enacting the Dancing Plague of 1518. But can an attacker convincingly spoof input and movement packets without any information about the peers? Or maybe there's a way to trick the server into revealing the peer information?

**Comment by Lejo1** [comment]
> The reason movement shouldn't be encrypted is because it's a low latency, high frequency kind of information which must be sent as quickly as possible, and there's nothing sensitive about it. Same goes for plain button inputs.

I disagree, you should either encrypt everything or nothing(except auth)
It's overcomplicated to not do it. You shouldn't notice the time needed for encryption when using a nice algorithm.

> I do wonder about the possibility of griefing a server with spoofed input packets, re-enacting the Dancing Plague of 1518. But can an attacker convincingly spoof input and movement packets without any information about the peers? Or maybe there's a way to trick the server into revealing the peer information?

Sure!
You just need to be somewhere between the server and client.
Than you could pretty much spoof anything.
You just don't get so easy between them ;)

I personally don't care if the connection is encrypted or not. Nobody would do such a hack to a minetest gaming server...

**Comment by hecktest** [comment]
Well I disagree with this disagreement, anything you do on the server, you have to multiply by the amount of players connected. The cost adds up quickly. If there is a trivial way to pass movement data as plaintext, do it.

**Comment by Lejo1** [comment]
Hmm, that's true.
Maybe the best is not to add more encryption and keep everything as it is.
Auth is encrypted that's the main thing.

Another idea is to encrypt only one channel(and put everything trusty in it)

**Comment by anon55555** [comment]
The encryption could just be done at a low level, everything in the UDP packets past the protocol id and source peer id could be encrypted (except auth because it already uses SRP and would be used for key agreement).

**Comment by red-001** [comment]
Badly implemented crypto might just provide a false sense of security, minetest being relatively small means not many eyes checking if the implementation is correct (for instance the SRP implementation had a bug in it that resulted in the salt not being random in some circumstances for ~4 years before anyone noticed). If you are concerned the risks of game chat, etc being intercepted you are likely better of informing people of relatively secure means of communication online (PGP, Signal Protocol, OTR, etc) and how to use them correctly. Public key cryptography might be useful for other purposes (maybe uniquely identifying servers for SPCSM even if the domain or port changes?) but I think it would be ill-advised to promote it as a protection against eavesdropping.

**Comment by appgurueu** [comment]
I suggest limiting the scope of this issue to what @hecktest proposed:

> Chat and form traffic should be encrypted. Maybe meta as well. Anything that involves big blobs of text.
> As it is, a MITM attack can be used to impersonate someone with privileges and send commands in their stead, including very dangerous ones like https://github.com/minetest/minetest/issues/10163, making this a vector for a server to get compromised. (We have enough botnets on the internet as it is.)

Chat is particularly important here.

**Comment by erlehmann** [comment]
I am in favor of full network encryption using DTLS or similar, but I want to add that encrypting chat end-to-end can be done with CSMs. As far as I know, the client Waspsalive contains an implementation of that.

**Comment by CLanguagePurist** [comment]
For network cryptography, you could do this as followed:

1. When initiating handshake it is best to use TLS or having a third party server to serves as a validator (either certificate authority or Serverlist server could do this) of this handshake session to avoid Trust on First Contact issue.
2. I generally suggest using Diffie-Hellman Elliptic Curve Cryptography on Curve25519 for public key exchange between the game server and the client. If we're on UDP instead of TCP, I would advise against ChaCha20 since it's a streaming protocol whereas UDP is stateless and can be arranged out of order which would pose a problem for streaming cipher. Instead, I would normally suggest either using AES, Threefish, or other symmetric block cipher algorithm (Just don't use 3DES please.)
3. You should only encrypt information that are sensitive like chat and authentication information, it doesn't make much sense to encrypt something like player movement however. You can instead use HMAC which is Hashed Message Authentication Code to ensure that packet is not tampered with and that it have a very low cost on computation requirement for such validation. HMAC can be summarized to simply hashing the content of the packet, encrypting the said hash and then append at the end of the packet.

Hope this helps.

**Comment by Desour** [comment]
I really doubt it makes any sense discussing about specific ciphers at this point...

----------------

What we need is:
-----------

(1) A way to authenticate servers.

The difficulties here are:
* Creating a LAN server should not be much more complicated than before.
* It should still be possible to host a (global) minetest server without a domain and/or webserver, because that would be a barrier that prevents some servers from actually using the secure networking.
* More?

To do this, we can generate X.509 certificates for the server.
We could add multiple possible ways (all possible, but optional) to trust it, e.g.:
* Via a webserver on the same domain.
* Via a checksum of the certificate, that the user copies from somewhere (i.e. a forum post).
* Via a checksum stored in the server favourites (only works for ensuring it's still the same server, and can become outdated).
* By displaying a checksum on first server join. On a LAN server, the user can then maybe just ask the hoster if the checksum is correct.
* Via some semi-central (that is, central, but the center can be configured, like the public server list) system.

---------

(2) A network protocol for secure channels.

The simplest here would be to just swap UDP with DTLS.

We could also optionally compile minetest with a QUIC implementation, with the datagram extension for unreliable packets.
In addition to security, this would also give us other networking benefits, like different priorities for different packets.

-------

(3) An addition to the minetest protocol to negotiate what protocol should be used, and to download the certificate.

This should happen before the normal minetest handshake happens.

Client and server both have a priority list for supported protocols (set via settings), i.e. `"encrypted_QUIC,partly_encrypted_DTLS,encrypted_DTLS,unencrypted_UDP"`.
The client can then before join download the supported protocols from the server, and chooses its favourite one. (This should ideally be optional for the server, to hide that there's a minetest server running.)
It also has to download and check the certificate.
We could additionally store the certificate and the supported protocols in the public server list.

**Comment by anon55555** [comment]
> If we're on UDP instead of TCP, I would advise against ChaCha20 since it's a streaming protocol whereas UDP is stateless and can be arranged out of order which would pose a problem for streaming cipher. Instead, I would normally suggest either using AES, Threefish, or other symmetric block cipher algorithm (Just don't use 3DES please.)

ChaCha20 is not a streaming protocol, it's a stream cipher which is no worse than a block cipher for encrypting packets. You can get multiple streams out of one key using different nonces (you need nonces either way unless you're using something like ECB mode).

> You should only encrypt information that are sensitive like chat and authentication information, it doesn't make much sense to encrypt something like player movement however.

It's simpler to just authenticate and encrypt everything rather than trying to decide what is and isn't sensitive. I doubt it would make a noticeable different to performance.

> You can instead use HMAC which is Hashed Message Authentication Code to ensure that packet is not tampered with and that it have a very low cost on computation requirement for such validation.

I'd prefer a MAC like Poly1305, which is faster and if packets are limited to 512 bytes it can be mathematically proven (assuming random keys are used) that an attacker making n forgery attempts has a success probability of at most n/2^98. This means, for example, to get even a one in a billion chance of successful forgery an attacker must send more than more than 4 zettabytes of authentication tags.

> HMAC can be summarized to simply hashing the content of the packet, encrypting the said hash and then append at the end of the packet.

That's not what HMAC does.

**Comment by CLanguagePurist** [comment]
I was referencing the HMAC code from memory when working with BouncyCastle which literally does this at the time. Also I have lost interest in using nor do I care to contribute to Minetest, so congratulation everyone.

**Comment by Emojigit** [comment]
I think we have to encrypt at least two things:
1. Chat messages, contains `/setpassword` and/or messages the government dislike.
2. [SSCSM](https://github.com/minetest/minetest/issues/7043), which require encryption to ensure safety.

**Comment by Desour** [comment]
There can also be confidential text in formspecs, formspec fields (filled out by client), mapblocks and whatnot.
It makes not much sense thinking about what we don't need to encrypt before benchmarking if there's any real performance penalty by just encrypting everything.

And SSCSM doesn't need encryption if we don't trust the server more than we already do.

**Comment by dr-bonez** [comment]
For what it's worth - using TLS also comes with the benefit of allowing virtual hosts using SNI. Allowing multiple services to listen on the same port at the same IP can be very useful for hosting solutions.

**Comment by HybridDog** [comment]
QUIC may be a good choice for Minetest:
* QUIC has useful features in addition to security: customizable congestion control, reliable and unreliable data transmission, prioritising of streams and perhaps even control of the latencies. These features are needed in Minetest, e.g. mapblock packets should be prioritised lower than packets which hurt or teleport a player. This is also an advantage over TLS.
* Efficient implementations of QUIC probably are or will be available. HTTP/3, which uses QUIC, has been standardized in 6 June 2022 and is used in the web, which could be a reason for a high demand for efficient QUIC.
* Minetest developers do not need to implement and maintain their own low-level protocol. 
* Implementing a custom secure protocol would be bad for security in general because there is no proof that the custom protocol is really secure and most Minetest developers are not security experts I think.
* It should be possible to get a shared key from SRP (the player's already secure login) and pass it to QUIC to establish the secure channel.


> I'd recommend against AES because without hardware acceleration (e.g. on mobile devices) it's slow and often vulnerable to cache-timing attacks (preventing them makes it slower).

I think Mobile devices which don't support browsing with a modern web browser are often too slow to play Minetest at all. I assume that modern web browsers always use AES for HTTPS (HTTP with TLS), so these devices should support AES well.

**Comment by anon55555** [comment]
> Implementing a custom secure protocol would be bad for security in general because there is no proof that the custom protocol is really secure and most Minetest developers are not security experts I think.

If we used ChaCha20 with Poly1305 to encrypt and authenticate raw UDP packets it wouldn't be too hard to prove security using indifferentiability:

In the *ideal world*, the client C and the server S are both is sending packets with distinct 64-bit nonces (these could replace Minetest RUDP's sequence numbers). After a packet is sent, the adversary A sees its size and nonce, and gains the capability to pass it on (possible repeatedly) to its intended recipient. The RUDP layer would stil need to be made robust against *malicious* packet dropping, delaying and replaying for security in this ideal world.

In the *real world*, the client C and the server S are both exchanging packets with the adversary B as specified by the protocol.

To prove security, construct a *simulator* Σ which simulates the capabilities of a real-world adversary using the capabilities of an ideal world adversary. This means the capabilities (Σ(A), C, S) and (B, C, S) are indistinguishable, which could be proven under the assumption that ChaCha20 is secure.

> I assume that modern web browsers always use AES for HTTPS (HTTP with TLS), so these devices should support AES well.

Your assumption is incorrect, see https://blog.cloudflare.com/do-the-chacha-better-mobile-performance-with-cryptography/.

**Comment by furkanmustafa** [comment]
This issue seems like stale for a while.

What about using WebRTC ?
* It uses TLS for handshake, and DTLS for UDP
* Can carry data packages (eg. game data)
* All of the cryptography decisions are offloaded to a standards workgroup
* Brings in easy to support browser networking as well
* Already established and battle tested libraries exist almost for all platforms

**Comment by red-001** [comment]
My view on this has changed in the past few years, I opened a draft PR to try and figure out how network encryption might look in Minetest #14970.

I've tested it locally and it works, but it's still very much a draft or proof of concept that would need more work.

**Comment by Darin755** [comment]
I have a few thoughts on a Minetest encrypted protocol. 

First off, the password authentication should be a different mechanism than the main encryption. It should involve the server sending some data and then the client signing the data with the password before sending it back. This would keep the password secure even if all the other security features are turned off or broken. It is totally acceptable (and recommended) to do the password check over a encrypted tunnel as well.

For the encryption itself I think it needs to have the following characteristics:
 - It should validate that the server has not been tapered with. This will likely require a server side certificate
 - It should use standard and well proven cryptography libraries. Such libraries should use modern ciphers and prohibit insecure ones even at the cost of breaking compatibility.
 - All data should be send encrypted by default. However, there should be an option to turn it off. There should not be an option to use weaker encryption or to partially encrypt traffic. (more options means more room for configuration error)
 - After the initial release of the encryption you should open it up to the community and eventually do a bug bounty for security vunerablities
 
Additionally, I do not think hardware that was made in the last 10 years will have trouble with cryptography. Modern CPUs have cryptographic accelerators that do all of the heavy lifting. Modern cryptographic libs are optimized heavily and do not add to much overhead. You shouldn't see any performance issues as there are a lot of factors that will be way more impactful.

For the protocol itself, you have two main libraries to choose from. 
 - [OpenSSL](https://www.openssl.org)
 - [libsodium](https://doc.libsodium.org)

Of those two I would go with OpenSSL. 

Lastly I would like to include the CIA Triad as that is always important to reference when talking about cybersecurity :

![The-Confidentiality-Integrity-Availability-CIA-triad-1707724215](https://github.com/user-attachments/assets/d3ad92e2-b6a5-4337-9977-d89aafd079a2) Credit: Research Gate

**Comment by furkanmustafa** [comment]
The points are quite good and on spot. But I think my comment (just 2 comments above) has been ignored too strongly.

Why bother designing and maintaining a cryptography stack and/or integration of it, as a game development team? It's a hard work, endlessly ongoing work, hard to get it right, easy to make mistakes. Why even make cryptography decisions?

There are standards, all necessary decisions already made, and being rediscussed and improved over time, by experts, and implementations of that standard are being maintained, by people who're focusing on that work. It's soooo valuable. Why not use those?

For example, this one; https://github.com/paullouisageneau/libdatachannel

Why choose between cryptography library? It's a platform/distro decision. And a platform/distro's responsibility to keep up-to-date and secure and stable.

A properly designed transport library (like above) will support most crypto libraries out of the box. So we don't have to bother thinking about it.

Please give it a thought. 🙏  The benefits are huge. The work is (hopefully a lot) less.

**Comment by appgurueu** [comment]
> First off, the password authentication should be a different mechanism than the main encryption. It should involve the server sending some data and then the client signing the data with the password before sending it back. This would keep the password secure even if all the other security features are turned off or broken. It is totally acceptable (and recommended) to do the password check over a encrypted tunnel as well.

Password authentication already uses [SRP](https://en.wikipedia.org/wiki/Secure_Remote_Password_protocol). The password is not sent over the wire when registering or logging into a server.

> It should use standard and well proven cryptography libraries. Such libraries should use modern ciphers and prohibit insecure ones even at the cost of breaking compatibility.

Minetest will not break network compatibility. If we implement support for encryption, new clients will try encrypted connections first, but servers will still support unencrypted connections for older clients, and new clients will still support unencrypted connections for older seervers.

Care needs to be taken to prevent downgrade attacks.

> After the initial release of the encryption you should open it up to the community

This will be "open to the community" at all points in time. Minetest is free and open source software. The development process is public. (Security advisories are published with a delay, but that's it.)

> and eventually do a bug bounty for security vunerablities

Sure, if you're funding it :)

**Comment by red-001** [comment]
> First off, the password authentication should be a different mechanism than the main encryption. It should involve the server sending some data and then the client signing the data with the password before sending it back. This would keep the password secure even if all the other security features are turned off or broken. It is totally acceptable (and recommended) to do the password check over a encrypted tunnel as well.

That would be a HMAC and it's a horrible way to try to implement password authentication. For one the server now needs to store password equivalent data (unless it's at least salted with a salt generated by the client), and there's no way to stop a malicious server from doing an active on-path attack to login as the user on another server. 

Minetest already uses SRP which avoids the first issue, and can be used in a way that avoids the second issue. Minetest currently doesn't use network encryption so it doesn't avoid it, and it's a bit difficult to do in a backwards compatible way. The cleanest way to do it that I found was to slightly modify the SRP protocol to include information about the underlying network encryption key. 

> All data should be send encrypted by default. However, there should be an option to turn it off. There should not be an option to use weaker encryption or to partially encrypt traffic. (more options means more room for configuration error)

Please try to learn something from TLS 1.2 and the mess that was. Secure systems should have as few configuration options as possible. One day we are going to get TLS 1.5 and it's probably only going to let you set the certificate you want to use for the server. There's no good reason to allow disabling traffic encryption if both sides support it. 

> A properly designed transport library (like above) will support most crypto libraries out of the box. So we don't have to bother thinking about it.

Expect you need to integrate that library in a backwards compatible way so that old clients can still connect over MTP/UDP. There's some talk of wanting to switch to a new networking backend, but that's been discussed for a long time.

**Comment by Darin755** [comment]
> > First off, the password authentication should be a different mechanism than the main encryption. It should involve the server sending some data and then the client signing the data with the password before sending it back. This would keep the password secure even if all the other security features are turned off or broken. It is totally acceptable (and recommended) to do the password check over a encrypted tunnel as well.
> 
> That would be a HMAC and it's a horrible way to try to implement password authentication. For one the server now needs to store password equivalent data (unless it's at least salted with a salt generated by the client), and there's no way to stop a malicious server from doing an active on-path attack to login as the user on another server.
> 
> Minetest already uses SRP which avoids the first issue, and can be used in a way that avoids the second issue. Minetest currently doesn't use network encryption so it doesn't avoid it, and it's a bit difficult to do in a backwards compatible way. The cleanest way to do it that I found was to slightly modify the SRP protocol to include information about the underlying network encryption key.
> 
> > All data should be send encrypted by default. However, there should be an option to turn it off. There should not be an option to use weaker encryption or to partially encrypt traffic. (more options means more room for configuration error)
> 
> Please try to learn something from TLS 1.2 and the mess that was. Secure systems should have as few configuration options as possible. One day we are going to get TLS 1.5 and it's probably only going to let you set the certificate you want to use for the server. There's no good reason to allow disabling traffic encryption if both sides support it.
> 
> > A properly designed transport library (like above) will support most crypto libraries out of the box. So we don't have to bother thinking about it.
> 
> Expect you need to integrate that library in a backwards compatible way so that old clients can still connect over MTP/UDP. There's some talk of wanting to switch to a new networking backend, but that's been discussed for a long time.

Thanks, I will look closer into SRP. I mostly proposed disabling encryption as some people in this issue stated that they are concerned about performance.

**Comment by Darin755** [comment]
I think the proper encryption to use is Datagram Transport layer security (dTLS)

https://en.wikipedia.org/wiki/Datagram_Transport_Layer_Security

There are a few libraries that support it and it seems pretty standard and secure. Theoretically, wouldn't this be as simple as just taking the network support functions and adding a wrapper around the encryption? I see the network code here https://github.com/minetest/minetest/tree/master/src/network. I am still trying to learn C++ and I have not finished looking though the code base. Maybe it would be possible to start a proof of concept? I am still very new at this so I am probably not much help but if someone with a C++ background wants to get started I will be happy to help if I can and I can test it.

**Comment by Darin755** [comment]
So is there consensus on going with QUIC or is it still up in the air? QUIC would be better from a public server perspective but it also means that a server operator probably would need to setup proper certs. QUIC has the downside of being almost to feature rich since it is really designed for the cell phone era. All of its multiplexing features aren't all that useful in the case of Luanti.

On the other hand you could use dTLS. dTLS is a bit rudimentary but the benefit is that it would be easy to add support. From a end user perspective you could prompt to trust like SSH does or you could simply rely on a user noticing that the signature changed. (The user could be shown a sha256 hash)

**Comment by sfan5** [comment]
This topic has not been discussed (most people's priorities are elsewhere) so there isn't a consensus.

**Comment by Darin755** [comment]
> This topic has not been discussed (most people's priorities are elsewhere) so there isn't a consensus.

@sfan5 Would there be any benefit to me working on a basic dTLS proof of concept? It should be fairly easy to add an additional layer to the network code and that would allow for performance testing. I get that there are higher priorities but I think network layer security is something that is important in 2025. I'm still new to C++ but I've done a bit of C programming and am familiar with OpenSSL so I think it should be possible for me to learn.

**Comment by appgurueu** [comment]
Personally I'm a fan of the idea of using QUIC, because it brings more benefits to the table than just encryption.
I'm not very worried about it being "too feature rich" as long as these features do not incur a significant cost for us.

I think we should decide on a protocol to move forward with - we do not want to first implement dTLS support, *then* implement QUIC later - and if I have to choose between QUIC and dTLS, the former seems like the better option.

**Comment by Desour** [comment]
I'm in favor of using QUIC eventually. But to do that you need the same auth stuff as for dtls, and IIRC some quic libs (ngtcp2) even use a tls connection to make it to quic or so. And when changing to QUIC, you'll have to replace the lower level part of the luanti protocol, not only the udp channels.

>QUIC would be better from a public server perspective but it also means that a server operator probably would need to setup proper certs.

That's no difference to dtls.
And there are other ways for auth, i.e. doing a key exchange and then authing with srp (see also #14970), or trust on first use.
(I wanted to try use a fake psk to do the key exchange under tls, but didn't find motivation to do it yet.)

**Comment by minlemon** [comment]
I am highly interested in seeing this through and will try to help where possible. I do wonder whatever happened to @red-001 and #14970?

**Comment by Darin755** [comment]
> I'm in favor of using QUIC eventually. But to do that you need the same auth stuff as for dtls, and IIRC some quic libs (ngtcp2) even use a tls connection to make it to quic or so. And when changing to QUIC, you'll have to replace the lower level part of the luanti protocol, not only the udp channels.

That's the part I probably can't help much with. I'm still new at this so digging into complex network code is more than I can realistically do. If there is something simpler to do I can help but doing advanced network programming is above my skill level.

I think network security is very important due to the growing popularity of Luanti especially when you consider that many users aren't aware that the lack of encryption could present serious hazards in some cases. Eventually someone will use Luanti to do something bad and it would better if spoofing and listening in on well known servers was harder. I'm sure I'm preaching to the choir but my point still stands.
